pkg/compressor/compressor.go

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package compressor
+
+import (
+	"compress/gzip"
+	"compress/lzw"
+	"compress/zlib"
+	"io"
+
+	"github.com/sirupsen/logrus"
+)
+
+// CompressSnapshots takes uncompressed data as input and compress the data according to Compression Policy
+// and write the compressed data into one end of pipe.
+func CompressSnapshots(data io.ReadCloser, compressionPolicy string) io.ReadCloser {
+	pReader, pWriter := io.Pipe()
+
+	var gWriter io.WriteCloser
+	logger := logrus.New().WithField("actor", "compressor")
+	logger.Infof("start compressing the snapshot using %v Compression Policy", compressionPolicy)
+
+	switch compressionPolicy {
+	case GzipCompressionPolicy:
+		gWriter = gzip.NewWriter(pWriter)
+
+	case LzwCompressionPolicy:
+		gWriter = lzw.NewWriter(pWriter, lzw.LSB, LzwLiteralWidth)
+
+	case ZlibCompressionPolicy:
+		gWriter = zlib.NewWriter(pWriter)
+
+	}
+
+	go func() {
+		defer pWriter.Close()
+		defer gWriter.Close()
+		defer data.Close()
+		n, err := io.Copy(gWriter, data)
+		if err != nil {
+			logger.Infof("compression failed: %v", err)
+			return
+		}
+		logger.Infof("compressed data size: %v KB", n/1024)
+	}()
+
+	return pReader
+}
+
+// DecompressSnapshots take compressed data and compressionPolicy as input and
+// it decompresses the data according to compression Policy and return uncompressed data.
+func DecompressSnapshots(data io.ReadCloser, compressionPolicy string) io.ReadCloser {
```

Comment:
I find to return the error difficult because [here](https://github.com/gardener/etcd-backup-restore/pull/293/files#diff-d5eff48d93149f39295c1872939affa38fd58bf430c468ce42dea93760e49253R87) lzw decompression is not returning the error

Reply:
Additional safety check never hurts :-)

Reply:
yeah ... I already handled the error for rest of 2 compression Policies and return the same data. you can  check [here](https://github.com/gardener/etcd-backup-restore/pull/293/files#diff-d5eff48d93149f39295c1872939affa38fd58bf430c468ce42dea93760e49253R75) and [here](https://github.com/gardener/etcd-backup-restore/pull/293/files#diff-d5eff48d93149f39295c1872939affa38fd58bf430c468ce42dea93760e49253R82)

Reply:
No. I meant returning and bubbling up the error and effectively failing the snapshot upload or restoration. I would not advice silently skipping compression/decompression. It is much better to fail early.

Reply:
I handled the error in `DecompressSnapshot()` but in case of `CompressSnapshot() `… error handling will be difficult because NewWriter() in all compression algos is not returning any error ….  we have to depend on error of  `n,err:= io.Copy()`  , so I  tried this but encountered some race condition when I run `make verify`, may be because `CompressSnapshot()` also have  thread running.